[LoongArch] Improve pattern matching for AddLike predicate

heiher opened the pull request:
This commit updates the pattern matching logic for the `AddLike` predicate in `LoongArchInstrInfo.td` to use the `isBaseWithConstantOffset` function provided by `CurDAG`. This optimization aims to improve the efficiency of pattern matching by identifying cases where the operation can be represented as a base address plus a constant offset, which can lead to more efficient code generation.

heiher committed:
[SelectionDAG] Improve isOrEquivalentToAdd

`SelectionDAGISel::isOrEquivalentToAdd` can only test operands that are
directly based on the stack frame. The improvement in this commit is to
add a call to `SelectionDAG::haveNoCommonBitsSet` for testing, which
helps optimize the code generation for `load (or base, imm)`.

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:
```diff
@@ -4214,7 +4214,7 @@ bool SelectionDAGISel::isOrEquivalentToAdd(const SDNode *N) const {
     return (Off >= 0) && (((A.value() - 1) & Off) == unsigned(Off));
   }
-  return false;
+  return CurDAG->haveNoCommonBitsSet(N->getOperand(0), N->getOperand(1));
```
I think we should migrate to isBaseWithConstantOffset and consider removing isOrEquivalentToAdd.

heiher replied:
Thanks. :+1:

heiher committed:
[LoongArch] Improve pattern matching for AddLike predicate

This commit updates the pattern matching logic for the `AddLike` predicate in
`LoongArchInstrInfo.td` to use the `isBaseWithConstantOffset` function provided
by `CurDAG`. This optimization aims to improve the effciency of pattern matching
by identifying cases where the operation can be represented as a base address
plus a constant offset, which can lead to more efficient code generation.

topperc reviewed (approved):
LGTM
